### PR TITLE
[lldb] Use LLVM's helper for Unicode conversion (NFC)

### PR DESCRIPTION
**Pull request by JDevlieghere**
The codecvt header has been deprecated in C++17. Use LLVM's unicode
helpers to convert between UTF-8 and UTF-16.

### CODE REVIEW

**Comment by JDevlieghere** [comment]
Looks like this is failing in CI. I'll take a look at this tomorrow. In the meantime I'd still like to merge #112446 and #112276.

File: lldb/source/Host/common/Editline.cpp [hunk 1]
```diff
@@ -76,6 +76,46 @@ using namespace lldb_private::line_editor;
...
+      std::use_facet<std::codecvt<wchar_t, char, std::mbstate_t>>(locale);
+
+  const size_t length = in.length();
+  std::string output(length + 1, 0x0);
+
+  std::mbstate_t mbs{};
+  const wchar_t *in_next;
+  char *out_next;
+
+  if (cvt.out(mbs, in.data(), in.data() + length + 1, in_next, output.data(),
```
Comment (labath) [inline lldb/source/Host/common/Editline.cpp:92]: in.begin/end() ?

File: lldb/source/Host/common/Editline.cpp [hunk 2]
```diff
@@ -76,6 +76,46 @@ using namespace lldb_private::line_editor;
...
 #endif // #if LLDB_EDITLINE_USE_WCHAR
 
+#if LLDB_EDITLINE_USE_WCHAR
+std::string ToBytes(const std::wstring &in) {
+  static std::locale locale("C.UTF-8");
+  static const auto &cvt =
+      std::use_facet<std::codecvt<wchar_t, char, std::mbstate_t>>(locale);
+
+  const size_t length = in.length();
+  std::string output(length + 1, 0x0);
```
Comment (labath) [inline lldb/source/Host/common/Editline.cpp:86]: This isn't right. A single wide character can take up to four bytes in the utf8 encoding. This works in `CompleteCharacter` as it's only interested in a single character (and it does the conversion in the other direction).  We might be able get away with overapproximating the size, given that the inputs here should be fairly small, but adding a loop with  partial conversions should not be that hard either.

File: lldb/source/Host/common/Editline.cpp [hunk 3]
```diff
@@ -76,6 +76,46 @@ using namespace lldb_private::line_editor;
...
+  return {};
+}
+
+std::wstring FromBytes(const std::string &in) {
+  static std::locale locale("C.UTF-8");
+  static const auto &cvt =
+      std::use_facet<std::codecvt<wchar_t, char, std::mbstate_t>>(locale);
+
+  const size_t length = in.length();
+  std::wstring output(length + 1, 0x0);
```
Comment (labath) [inline lldb/source/Host/common/Editline.cpp:105]: I believe this direction should be ok (even without the spurious +1).

**Review by labath — COMMENTED** [review]
If you can't reproduce the buildbot errors, I can take a look, but I think the failures are likely due to the string length problems.

**Comment by labath** [comment]
Actually (and, in retrospect, obviously) llvm already has [functions](https://github.com/llvm/llvm-project/blob/7dc2542ac24fcae89dfd179fa58c4ec4fb959e2b/llvm/include/llvm/Support/ConvertUTF.h#L227C6-L227C23) for wstring<->string conversions, so it would be best to use those.

**Comment by JDevlieghere** [comment]
@labath FYI this is ready for review.

**Comment by JDevlieghere** [comment]
Friendly ping

File: lldb/source/Host/common/Editline.cpp [hunk 4]
```diff
@@ -1574,25 +1584,23 @@ bool Editline::CompleteCharacter(char ch, EditLineGetCharType &out) {
...
-    std::mbstate_t state = std::mbstate_t();
     input.push_back(ch);
-    switch (cvt.in(state, input.begin(), input.end(), from_next, &out, &out + 1,
-                   to_next)) {
-    case std::codecvt_base::ok:
+    const char *cur_ptr = input.begin();
+    const char *end_ptr = input.end();
+    llvm::UTF32 code_point = 0;
+    llvm::ConversionResult cr = llvm::convertUTF8Sequence(
+        (const llvm::UTF8 **)&cur_ptr, (const llvm::UTF8 *)end_ptr, &code_point,
```
Comment (labath) [inline lldb/source/Host/common/Editline.cpp:1594]:
```suggestion
    auto *cur_ptr = reinterpret_cast<const llvm::UTF8 *>(input.begin());
    auto *end_ptr = reinterpret_cast<const llvm::UTF8 *>(input.end());
    llvm::UTF32 code_point = 0;
    llvm::ConversionResult cr = llvm::convertUTF8Sequence(
        &cur_ptr, end_ptr, &code_point,
```

I think this is "less undefined" :P

**Review by labath — APPROVED** [review]

**Pull request merged**